pkg/cosign/fetch.go

```diff
@@ -138,20 +138,42 @@ func FetchAttestationsForReference(ctx context.Context, ref name.Reference, opts
 		return nil, fmt.Errorf("no attestations associated with %s", ref)
 	}
 
-	attestations := make([]AttestationPayload, len(l))
+	attestations := make([]AttestationPayload, 0, len(l))
 	var g errgroup.Group
 	g.SetLimit(runtime.NumCPU())
-	for i, att := range l {
-		i, att := i, att
+
+	for _, att := range l {
+		if predicateType != "" {
+			anns, err := att.Annotations()
+			if err != nil {
+				return nil, err
+			}
+			pt, ok := anns["predicateType"]
+			// Skip attestation if predicateType annotation is not present or predicateType annotation does not match supplied predicate type
+			if !ok || pt != predicateType {
+				continue
+			}
+		}
+		att := att
+		var a AttestationPayload
 		g.Go(func() error {
 			attestPayload, _ := att.Payload()
-			return json.Unmarshal(attestPayload, &attestations[i])
+			err := json.Unmarshal(attestPayload, &a)
```

Comment:
Question: Wouldn't it be an alternative to extract the predicateType from the attestation payload instead of relying on annotations ?

Reply:
It would be, but doing so requires decoding each payload first to inspect the predicateType, which is both more memory-intensive and mean that third-party tools (written in languages other than Go) that are trying to find the desired attestation must download each attestation blob and decode the payload as well.  

Adding the annotation means the desired attestation layer can be found using basic `GET` requests against the docker v2 API, which means other tools can integrate more easily with the artifacts produced by Cosign.

Reply:
Yes, that makes sense although this command will be misleading as the flag describes an action that could not happen. If the command flag aims to download attestations with a predicate type, if there aren't annotations for these attestations it will return an empty list even if there are attestations.

Reply:
Ah good point, I will update the help text and add unit tests this week.  I think the behavior makes sense though - if the flag is provided and no *annotations* are matched, throw an error to avoid confusion around why no attestations were returned, else if no flag was provided, return all attestations.

Reply:
Well, on second thought - this MR will ensure that the `predicateType` annotation *always* exists.  So the behavior would be the same.  The only exception to this is legacy attestations put in place before this PR.  Either way, the text should be clarified.